ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryMonitor.java

```diff
@@ -0,0 +1,215 @@
+          final QueryId queryId = query.getQueryId();
+          final KafkaStreams.State queryState = query.getState();
+
+          // If the query was restarted previously, check if it needs another restart; or if the
+          // query is now up and running, then remove it from the restart list.
+          if (queriesRetries.containsKey(queryId)) {
+            final RetryEvent retryEvent = queriesRetries.get(queryId);
+            if (ticker.read() > retryEvent.nextRestartTimeMs()) {
+              if (queryState == KafkaStreams.State.ERROR) {
+                // Retry again if it's still in ERROR state
+                retryEvent.restart();
+              } else {
+                // Query is not in ERROR state anymore.
+                queriesRetries.remove(queryId);
+              }
+            }
+          } else if (queryState == KafkaStreams.State.ERROR) {
+            // Restart new query in ERROR state, and add it to the list of retries.
+            final RetryEvent retryEvent = new RetryEvent(
+                ksqlEngine, queryId, BASE_WAITING_TIME_MS, retryBackoffMaxMs, ticker);
+
+            queriesRetries.put(queryId, retryEvent);
+            retryEvent.restart();
+          }
+        });
+  }
+
+  private class Runner implements Runnable {
+    @Override
+    public void run() {
+      LOG.info("KSQL query monitor started.");
+
+      while (!closed) {
+        try {
+          restartFailedQueries();
+        } catch (final Exception e) {
+          LOG.warn("KSQL query monitor found an error attempting to restart failed queries.", e);
+        }
+      }
+    }
+  }
+
+  static class RetryEvent {
+    private final KsqlEngine ksqlEngine;
+    private final QueryId queryId;
+    private final Ticker ticker;
+
+    private int numRetries = 0;
+    private long waitingTimeMs;
+    private long expiryTimeMs;
+    private long retryBackoffMaxMs;
+    private long baseWaitingTimeMs;
+
+    RetryEvent(
+        final KsqlEngine ksqlEngine,
+        final QueryId queryId,
+        final long baseWaitingTimeMs,
+        final long retryBackoffMaxMs,
+        final Ticker ticker
+    ) {
+      this.ksqlEngine = ksqlEngine;
+      this.queryId = queryId;
+      this.ticker = ticker;
+
+      this.baseWaitingTimeMs = baseWaitingTimeMs;
+      this.waitingTimeMs = baseWaitingTimeMs;
+      this.expiryTimeMs = ticker.read() + baseWaitingTimeMs;
+      this.retryBackoffMaxMs = retryBackoffMaxMs;
+    }
+
+    public long nextRestartTimeMs() {
+      return expiryTimeMs;
+    }
+
+    public int getNumRetries() {
+      return numRetries;
+    }
+
+    public void restart() {
+      numRetries++;
+
+      LOG.info("Restarting query {} (attempt #{})", queryId, numRetries);
+
+      // Stop the queryId using the current QueryMetadata
+      ksqlEngine.getPersistentQuery(queryId).ifPresent(q -> q.stop());
+
+      // Reset the internal KafkaStreams query. This creates a new QueryMetadata.
+      ksqlEngine.resetQuery(queryId);
```

Comment:
what happens if we fail after closing the kafka streams query but before we call reset query? then the kafka streams application will be `DEAD` or `PENDING_SHUTDOWN` and we wouldn't try restarting it again. I think we should make sure that this block is transactional - either all of it happens, or none of it happens. That might be a challenge, but I think it's important